actors/migration/reward.go

```diff
@@ -3,35 +3,153 @@ package migration
+
+	// Set effective network power to the value it would be at the end of the
+	// prior epoch if the current parameters had been in place at network start
+	cumSumRealized := big.Lsh(inState.CumsumRealized, math2.Precision128) // Q.0 => Q.128
+	acc0 := big.Mul(cumSumRealized, gTauOverB0)                           // Q.128 * Q.128 => Q.256
+	acc0 = big.Rsh(acc0, math2.Precision128)                              // Q.256 => Q.128
+	acc0 = big.Add(acc0, one)
+	acc0 = math2.Ln(acc0)
+	acc0 = big.Mul(acc0, oneOverGTau)                               // Q.128 * Q.128 => Q.256
+	postUpgradeTheta := big.Rsh(acc0, math2.Precision128)           // Q.256 => Q.128
+	effNetworkTime := big.Rsh(postUpgradeTheta, math2.Precision128) // Q.128 => Q.0
+	effNetworkTime = big.Add(effNetworkTime, big.NewInt(1))
+	if !effNetworkTime.Int.IsInt64() {
+		return cid.Undef, big.Zero(), xerrors.Errorf("new effective network time chain epoch out of bounds %v", effNetworkTime)
+	}
+	outEffectiveNetworkTime := abi.ChainEpoch(effNetworkTime.Int.Int64())
+
+	// Set effective baseline power given the new effective network time and
+	// baseline function.
+	outEffectiveBaselinePower := math2.ExpBySquaring(reward2.BaselineExponent, int64(outEffectiveNetworkTime)) // Q.128
+	outEffectiveBaselinePower128 := big.Mul(reward2.BaselineInitialValue, outEffectiveBaselinePower)           // Q.0 * Q.128 => Q.128
+	outEffectiveBaselinePower = big.Rsh(outEffectiveBaselinePower128, math2.Precision128)                      // Q.128 => Q.0
+
+	// Set cumsum baseline given new effective network time and baseline function
+	outCumSumBaseline := big.Sub(outEffectiveBaselinePower128, big.Lsh(reward2.BaselineInitialValue, math2.Precision128)) // Q.128 - Q.128 => Q.128
+	outCumSumBaseline = big.Mul(outCumSumBaseline, oneOverGTau)                                                           // Q.128 * Q.128 => Q.256
+	outCumSumBaseline = big.Rsh(outCumSumBaseline, 2*math2.Precision128)                                                  // Q.256 => Q.0
+
+	// Set reward filter postiion and velocity values
+
+	// Filter position is set by evaluating the closed form expression for expected
+	// block reward at a particular epoch assuming that the new baseline function
+	// had been in place at network start
+	acc1 := big.Mul(reward2.BaselineInitialValue, oneOverGTau) // Q.0 * Q.128 => Q.128
+	acc1 = big.Add(acc1, big.Lsh(inState.CumsumRealized, math2.Precision128))
+	rawBytePower256 := big.Lsh(thisEpochRawBytePower, 2*math2.Precision128) // Q.0 => Q.256
+	acc1 = big.Div(rawBytePower256, acc1)                                   // Q.256 / Q.128 => Q.128
+	acc1 = big.Add(acc1, one)
+	acc1 = math2.Ln(acc1)
+	acc1 = big.Lsh(acc1, math2.Precision128)                         // Q.128 => Q.256
+	acc1 = big.Div(acc1, big.Lsh(big.NewInt(6), math2.Precision128)) // Q.256 / Q.128 => Q.128
+	acc1 = big.NewFromGo(math2.ExpNeg(acc1.Int))
```

Comment:
I'm not sure what `NewFromGo` is doing here - is it just converting between two copies of the bigint library?

Reply:
Yes its from an inconsistency between using raw go big.Int's and the actors wrapped big.Int type in expneg and polyval.  I came close to cleaning this up but it fell just below the threshold of feeling worth it.